loopy/transform/loop_fusion.py

```diff
@@ -0,0 +1,801 @@
+            # is if "l" is nested within "i,j"-loops.
+            self.tree = _add_inner_loops(self.tree, outer_loop, not_seen_inames)
+
+        # {{{ update iname to node id
+
+        for iname in outer_loop:
+            self.iname_to_tree_node_id = self.iname_to_tree_node_id.set(iname,
+                                                                        outer_loop)
+
+        if inner_loop is not None:
+            for iname in inner_loop:
+                self.iname_to_tree_node_id = self.iname_to_tree_node_id.set(
+                    iname, inner_loop)
+
+        for iname in not_seen_inames:
+            self.iname_to_tree_node_id = self.iname_to_tree_node_id.set(
+                iname, not_seen_inames)
+
+        # }}}
+
+        # }}}
+
+        assert not (outer_redn_inames & redn_inames)
+        return super().map_reduction(
+            expr,
+            expn_state,
+            outer_redn_inames=(outer_redn_inames | redn_inames))
+
+
+def _add_reduction_loops_in_partial_loop_nest_tree(kernel, tree):
+    """
+    Returns a partial loop nest tree with the loop nests corresponding to the
+    reduction inames added to *tree*.
+    """
+    from loopy.symbolic import SubstitutionRuleMappingContext
+    rule_mapping_context = SubstitutionRuleMappingContext(
+        kernel.substitutions, kernel.get_var_name_generator())
+    reduction_loop_inserter = ReductionLoopInserter(rule_mapping_context, tree)
+
+    def does_insn_have_reduce(kernel, insn, *args):
+        return bool(insn.reduction_inames())
+
+    reduction_loop_inserter.map_kernel(kernel,
+                                       within=does_insn_have_reduce,
+                                       map_tvs=False, map_args=False)
+    return reduction_loop_inserter.tree
+
+
+def _get_partial_loop_nest_tree_for_fusion(kernel):
+    from loopy.schedule.tools import _get_partial_loop_nest_tree
+    tree = _get_partial_loop_nest_tree(kernel)
+    tree = _add_reduction_loops_in_partial_loop_nest_tree(kernel, tree)
+    return tree
+
+
+def get_kennedy_unweighted_fusion_candidates(kernel: LoopKernel,
+                                             candidates: FrozenSet[str],
+                                             prefix="ifused"):
+    """
+    Returns the fusion candidates mapping that could be fed to
+    :func:`rename_inames_in_batch` similar to Ken Kennedy's Unweighted
+    Loop-Fusion Algorithm.
+
+    .. attribute:: prefix
+
+        Prefix for the fused inames.
+    """
+    from loopy.kernel.data import ConcurrentTag
+    from loopy.schedule.tools import (
+        _get_iname_to_tree_node_id_from_partial_loop_nest_tree)
+    from collections.abc import Collection
+    assert not isinstance(candidates, str)
+    assert isinstance(candidates, Collection)
+    assert isinstance(kernel, LoopKernel)
+
+    candidates = frozenset(candidates)
+    vng = kernel.get_var_name_generator()
+    fused_chunks = {}
+
```

Comment:
Note to self: Add a sanity check here if `candidates` are not a part of `kernel`.